dpnp/dpnp_utils/dpnp_utils_linearalgebra.py

```diff
@@ -2010,24 +2028,24 @@ def dpnp_matmul(
     elif x1.size == 0 or x2.size == 0:
         result.fill(0)
     else:
-        # input arrays should have the proper data type
-        # and be C_CONTIGUOUS or F_CONTIGUOUS
+        # input arrays should have the proper data type and
+        # their base (last 2-dimensions) to be c-contiguous or f-contiguous
         dep_events_list = []
         host_tasks_list = []
-        contig_copy = not (x1.flags.c_contiguous or x1.flags.f_contiguous)
+        contig_flag = _define_contig_flag(x1)
```

Comment:
The failing `tests/third_party/cupy/linalg_tests/test_product.py::TestDot_param_24` test shows that we might have an unexpected memory copy:
```python
shape_a, shape_b = (2, 1), (1, 3)
a = dpnp.ones(shape_a[::-1]).T
b = dpnp.ones(shape_b[::-1]).T

dpnp.dot(a, b)    # which is leveraging on dpnp.matmul(a, b)
```
there will be new strides returned during reshape at lines#1699: `(1, 0)`, when the input one was `(1, 1)`.
So, `_define_contig_flag` will return `False` instead of expecting `True` value and it will cause the extra memory copy below in `_copy_array`.
I've filed an issue [gh-1664](https://github.com/IntelPython/dpctl/issues/1664) to clarify if it is expected `reshape` behavior. Otherwise (if it's expected) then we will need to update the implementation here and to consider that use case.

Reply:
Regardless of dpctl issue, in future PR, we will implement alternative approaches for similar cases. For instance, we can fall backs on OneMKL dot function (similar to NumPy) when input arrays are inherently 1D.
```
>>> a=numpy.ones((1,2))
>>> b=numpy.ones((2,1))
>>> numpy.matmul(a,b)
MKL_VERBOSE DDOT(2,0x55a4609797f0,1,0x55a46097cd60,1) 246.44us CNR:OFF Dyn:1 FastMM:1 TID:0  NThr:6
 ```
 
Or for such a case, we do not need to call any functions from OneMKL backend (NumPy does not), we can just call `dpnp.kron`
```
>>> c=numpy.ones((2,1))
>>> d=numpy.ones((1,3))
>>> numpy.matmul(c,d)
array([[1., 1., 1.],
       [1., 1., 1.]])
```